transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,444 @@
+            null,
+            null,
+            null)) {
+      while (cursor.moveToNext()) {
+        long id = cursor.getLong(0);
+        EventInternal event =
+            EventInternal.builder()
+                .setTransportName(cursor.getString(1))
+                .setPriority(toPriority(cursor.getInt(2)))
+                .setEventMillis(cursor.getLong(3))
+                .setUptimeMillis(cursor.getLong(4))
+                .setPayload(cursor.getBlob(5))
+                .build();
+        events.add(PersistedEvent.create(id, backendName, event));
+      }
+    }
+    return events;
+  }
+
+  /** Loads metadata pairs for given events. */
+  private Map<Long, Set<Metadata>> loadMetadata(SQLiteDatabase db, List<PersistedEvent> events) {
+    Map<Long, Set<Metadata>> metadataIndex = new HashMap<>();
+    StringBuilder whereClause = new StringBuilder("event_id IN (");
+    for (int i = 0; i < events.size(); i++) {
+      whereClause.append(events.get(i).getId());
+      if (i < events.size() - 1) {
+        whereClause.append(',');
+      }
+    }
+    whereClause.append(')');
+
+    try (Cursor cursor =
+        db.query(
+            "event_metadata",
+            new String[] {"event_id", "name", "value"},
+            whereClause.toString(),
+            null,
+            null,
+            null,
+            null)) {
+      while (cursor.moveToNext()) {
+        long eventId = cursor.getLong(0);
+        Set<Metadata> currentSet = metadataIndex.get(eventId);
+        if (currentSet == null) {
+          currentSet = new HashSet<>();
+          metadataIndex.put(eventId, currentSet);
+        }
+        currentSet.add(new Metadata(cursor.getString(1), cursor.getString(2)));
+      }
+    }
+    return metadataIndex;
+  }
+
+  /** Populate metadata into the events. */
+  private List<PersistedEvent> join(
+      List<PersistedEvent> events, Map<Long, Set<Metadata>> metadataIndex) {
+    ListIterator<PersistedEvent> iterator = events.listIterator();
+    while (iterator.hasNext()) {
+      PersistedEvent current = iterator.next();
+      if (!metadataIndex.containsKey(current.getId())) {
+        continue;
+      }
+      EventInternal.Builder newEvent = current.getEvent().toBuilder();
+
+      for (Metadata metadata : metadataIndex.get(current.getId())) {
+        newEvent.addMetadata(metadata.key, metadata.value);
+      }
+      iterator.set(
+          PersistedEvent.create(current.getId(), current.getBackendName(), newEvent.build()));
+    }
+    return events;
+  }
+
+  private <T> T retryIfDbLocked(
+      long lockTimeoutMs, Producer<T> retriable, Function<Throwable, T> failureHandler) {
+    long startTime = monotonicClock.getTime();
+    do {
+      try {
+        return retriable.produce();
+      } catch (SQLiteDatabaseLockedException ex) {
+        if (monotonicClock.getTime() >= startTime + lockTimeoutMs) {
+          return failureHandler.apply(ex);
+        }
+        SystemClock.sleep(LOCK_RETRY_BACK_OFF);
```

Comment:
In practice, will there be different threads competing for this?
Does it have to be a read-write lock or do transactional writes make concurrent reads thread safe?

Reply:
There won't be as all db access will happen on one dedicated thread. We are also intentionally not using WAL so concurrent writes are not allowed